hardware_interface/include/hardware_interface/sensor.hpp

```diff
@@ -0,0 +1,204 @@
+   * can be recived by get_state_interfaces() function.
+   *
+   * \param state list of doubles with states of the hardware.
+   */
+  HARDWARE_INTERFACE_EXPORT
+  void get_complete_state(std::vector<double> & state) const
+  {
+    state.clear();
+    for (const auto & internal_state : states_) {
+      state.push_back(internal_state);
+    }
+  }
+
+  /**
+   * \brief Set state list for the sensor. This function is used by the hardware to set its actual
+   * state. The parameters state, and interfaces have the same order and number of elements. Using
+   * the interfaces list, the hardware can choose which values to set.
+   *
+   * \param state list of doubles with states of the hardware.
+   * \param interfaces list of interfaces on which states have to be provided.
+   * \return return_type::INTERFACE_VALUE_SIZE_NOT_EQUAL if state and interfaces arguments do not
+   * have the same length; return_type::INTERFACE_NOT_FOUND if one of provided interfaces is not
+   * defined for the sensor; return_type::OK otherwise.
+   */
+  HARDWARE_INTERFACE_EXPORT
+  return_type set_state(
+    const std::vector<double> & state,
+    const std::vector<std::string> & interfaces)
+  {
+    if (state.size() != interfaces.size()) {
+      return return_type::INTERFACE_VALUE_SIZE_NOT_EQUAL;
+    }
+    return_type ret = return_type::OK;
+    bool found;
+
+    for (uint i = 0; i < info_.state_interfaces.size(); i++) {
+      found = false;
+      for (uint j = 0; j < info_.state_interfaces.size(); j++) {
+        if (!interfaces[i].compare(info_.state_interfaces[j])) {
+          states_[j] = state[i];
+          found = true;
+          break;
+        }
+      }
+      if (!found) {
+        ret = return_type::INTERFACE_NOT_FOUND;
+        break;
+      }
+    }
+    return ret;
+  }
+
+  /**
+   * \brief Set complete state list from the sensor.This function is used by the hardware to set its
+   * complete actual state. The state values have the same order as interfaces which can be recived
+   * by get_state_interfaces() function.
+   *
+   * \param state list of doubles with states from the hardware.
+   * \return return_type::INTERFACE_VALUE_SIZE_NOT_EQUAL is command size is not equal to number of
+   * joint's state interfaces, return_type::OK otherwise.
+   */
+  HARDWARE_INTERFACE_EXPORT
+  return_type set_complete_state(const std::vector<double> & state)
+  {
+    if (state.size() == states_.size()) {
+      for (uint i = 0; i < states_.size(); i++) {
+        states_[i] = state[i];
+      }
+    } else {
+      return return_type::INTERFACE_VALUE_SIZE_NOT_EQUAL;
+    }
+    return return_type::OK;
+  }
+
+protected:
+  ComponentInfo info_;
+  std::vector<double> states_;
+
+  /**
+   * \brief Configure base sensor class based on the description in the robot's URDF file.
+   *
+   * \param joint_info structure with data from URDF.
+   * \return return_type::OK
+   */
+  return_type configure_base(const ComponentInfo & joint_info)
```

Comment:
this should just be a regular `configure` implementation (public). The user then might or might not override this function.